nbt/src/main/java/net/kyori/adventure/nbt/BinaryTagWriterImpl.java

```diff
@@ -0,0 +1,92 @@
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package net.kyori.adventure.nbt;
+
+import java.io.BufferedOutputStream;
+import java.io.DataOutput;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Map;
+import java.util.zip.GZIPOutputStream;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import static net.kyori.adventure.nbt.IOStreamUtil.closeShield;
+
+final class BinaryTagWriterImpl implements BinaryTagIO.Writer {
+  static final BinaryTagIO.Writer INSTANCE = new BinaryTagWriterImpl();
+
+  @Override
+  public void write(final @NonNull CompoundBinaryTag tag, final @NonNull Path path, final BinaryTagIO.@Nullable Compression compression) throws IOException {
+    try(final OutputStream os = Files.newOutputStream(path)) {
+      this.write(tag, os, compression);
+    }
+  }
+
+  @Override
+  public void write(final @NonNull CompoundBinaryTag tag, @NonNull OutputStream output, final BinaryTagIO.@Nullable Compression compression) throws IOException {
+    output = closeShield(output);
+    if(compression != null) {
+      output = compression.compress(output);
+    }
+    try(final DataOutputStream dos = new DataOutputStream(new BufferedOutputStream(output))) {
+      this.write(tag, (DataOutput) dos);
+    }
+  }
+
+  @Override
+  public void write(final @NonNull CompoundBinaryTag tag, final @NonNull DataOutput output) throws IOException {
+    output.writeByte(BinaryTagTypes.COMPOUND.id());
+    output.writeUTF(""); // write empty name
+    BinaryTagTypes.COMPOUND.write(tag, output);
+  }
+
+  @Override
+  public void writeNamed(final Map.@NonNull Entry<String, CompoundBinaryTag> tag, final @NonNull Path path, final BinaryTagIO.@Nullable Compression compression) throws IOException {
+    try(final OutputStream os = Files.newOutputStream(path)) {
+      this.writeNamed(tag, os, compression);
+    }
+  }
+
+  @Override
+  public void writeNamed(final Map.@NonNull Entry<String, CompoundBinaryTag> tag, @NonNull OutputStream output, final BinaryTagIO.@Nullable Compression compression) throws IOException {
+    output = closeShield(output);
+    if(compression != null) {
+      output = compression.compress(output);
+    }
+    try(final DataOutputStream dos = new DataOutputStream(new BufferedOutputStream(new GZIPOutputStream(closeShield(output))))) {
```

Comment:
bug here with `GZIPOutputStream` always applied, you might want to add some tests to cover this :P